### PR TITLE
Ability to register metrics collectors

### LINKED ISSUE #203: The ability to register metric collectors from application
Need a new RPC method to register metric from the application, metrics must only be declared once so consecutive runs would have no effect.

### PR DESCRIPTION
**Pull request by 48d90782**
This is Draft PR, still in progress
Resolves #203

### CODE REVIEW

File: service/metrics/config.go [hunk 1]
```diff
@@ -15,6 +15,31 @@ type Config struct {
 	Collect map[string]Collector
 }
 
+type NamedCollector struct {
```
Comment (wolfy-j) [inline service/metrics/config.go]: need json aliases here since it will be passed from PHP

Comment (48d90782) [reply service/metrics/config.go]: Done

File: .gitignore [hunk 2]
```diff
@@ -1,6 +1,7 @@
-.idea/*
+.idea
 composer.lock
-vendor/
+vendor
 builds/
 tests/vendor/
-go.sum
\ No newline at end of file
+.rr.yaml
```
Comment (wolfy-j) [inline .gitignore]: why is this file added, its default template which is used to init config in some apps, leaving it out of the repository won't break anything but will have new users ux

File: cmd/util/exit.go [hunk 3]
```diff
@@ -6,6 +6,10 @@ import (
 
 // ExitWithError prints error and exits with error code`.
 func ExitWithError(err error) {
-	Panicf("<red+hb>Error:</reset> <red>%s</reset>\n", err)
+	errP := Panicf("<red+hb>Error:</reset> <red>%s</reset>\n", err)
+	if errP != nil {
+		// in case of error during Panicf, print this error via build-int print function
+		println("error occurred during fmt.Fprint: " + err.Error())
+	}
```
Comment (wolfy-j) [inline cmd/util/exit.go]: thx

File: go.mod [hunk 4]
```diff
@@ -1,31 +1,24 @@
...
+	github.com/prometheus/client_golang v1.4.1
+	github.com/shirou/gopsutil v2.20.1+incompatible
+	github.com/sirupsen/logrus v1.4.2
+	github.com/spf13/cobra v0.0.5
+	github.com/spf13/viper v1.6.2
+	github.com/spiral/goridge v2.1.4+incompatible
+	github.com/stretchr/testify v1.4.0
+	github.com/yookoala/gofast v0.4.0
+	golang.org/x/net v0.0.0-20200202094626-16171245cfb2
 )
```
Comment (wolfy-j) [inline go.mod]: nice cleanup! is it legal?

File: service/metrics/config.go [hunk 5]
```diff
@@ -15,25 +15,45 @@ type Config struct {
...
+	// Name of the collector
+	Name string `json:"name"`
+
+	// Collector structure
+	Collector `json:"collector"`
```
Comment (wolfy-j) [inline service/metrics/config.go]: Not sure I remember how JSON annotation works on an inherited struct, does it work at or all or it's redundant?

File: service/metrics/config.go [hunk 6]
```diff
@@ -15,25 +15,45 @@ type Config struct {
...
+	// Gauge type
+	Gauge CollectorType = "gauge"
+
+	// Counter type
+	Counter CollectorType = "counter"
+
+	// Summary type
+	Summary CollectorType = "summary"
+)
+
```
Comment (wolfy-j) [inline service/metrics/config.go]: thx

File: service/metrics/rpc.go [hunk 7]
```diff
@@ -126,6 +130,98 @@ func (rpc *rpcServer) Observe(m *Metric, ok *bool) (err error) {
...
+	*ok = true
+	return nil
+}
+// Register is used to register new collector in prometheus
+// THE TYPES ARE:
+// 	NamedCollector -> Collector with the name
+// 	bool -> RPC reply value
+// RETURNS:
+// 	error
+func (rpc *rpcServer) Register(c *NamedCollector, ok *bool) (err error) {
```
Comment (wolfy-j) [inline service/metrics/rpc.go]: Can you rename it to `Declare` as convention for soft inserts?

File: .gitignore (hunk 2 above)
Comment (48d90782) [reply .gitignore]: Default config already committed into the repo. Since .rr.yaml is just a sample of config there are 2 options:
1. Put this file into .gitignore and allow users to make their own local changes in file
2. Remove it from .gitignore, but in this way, I have to create a copy of that file for local testing for example.
So, which variant would you prefer?

File: go.mod (hunk 4 above)
Comment (48d90782) [reply go.mod]: I am not sure :)

File: .gitignore (hunk 2 above)
Comment (wolfy-j) [reply .gitignore]: 1.

File: service/metrics/config.go (hunk 5 above)
Comment (48d90782) [reply service/metrics/config.go]: yeah, It's not redundant. Under `collector` would be inherited structure with json tagged fields

File: service/metrics/rpc.go (hunk 7 above)
Comment (48d90782) [reply service/metrics/rpc.go]: Sure
